File: drills/elixir/solutions/nc140_same_tree.exs
defmodule Solution do
  # Walk both trees in step. Two empties match, an empty and a node never do,
  # and two nodes match when their values do and both pairs of children do.
  # Elixir's structural equality would answer this in one character -- the point
  # of writing it out is that the same shape is what Subtree of Another Tree and
  # Symmetric Tree are built from.
  def is_same_tree(nil, nil), do: true

  def is_same_tree({a, a_left, a_right}, {b, b_left, b_right}) do
    a == b and is_same_tree(a_left, b_left) and is_same_tree(a_right, b_right)
  end

  def is_same_tree(_first, _second), do: false
end

File: drills/go/harnesses/nc140_same_tree.go
package main

func main() {
	run(func() []testCase {
		return []testCase{
			tc("isSameTree([1,2,3], [1,2,3])", true, isSameTree(tree(1, 2, 3), tree(1, 2, 3))),
			tc("isSameTree([1,2], [1,null,2])", false, isSameTree(tree(1, 2), tree(1, x, 2))),
			tc("isSameTree([1,2,1], [1,1,2])", false, isSameTree(tree(1, 2, 1), tree(1, 1, 2))),
			tc("isSameTree([], [])", true, isSameTree(tree(), tree())),
			tc("isSameTree([1], [])", false, isSameTree(tree(1), tree())),
		}
	})
}
